File: capacitor/plugins-library/ios/Plugin/DataStorage/DataStorage.swift
import Foundation
import Capacitor

@objc(DataStorage)
public class DataStorage: CAPPlugin {
    
    private var dbName = "datastorage.sqlite"
    
    @objc func database(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.error(DataStorageError.EmptyDatabaseName)
            return
        }
        self.dbName = "\(name).sqlite"
        call.success([:])
    }
    
    @objc func delete(_ call: CAPPluginCall) {
        guard let table = call.getString("table") else {
            call.error(DataStorageError.EmptyTable)
            return
        }
        
        guard let key = call.getString("key") else {
            call.error(DataStorageError.EmptyKey)
            return
        }
        
        let db = SqliteDB(dbPath: self.dbName, tableName: table)
        
        if !db.existDatabase() {
            call.error(DataStorageError.DatabaseNotFound)
            return
        }
        
        if let errorMessage = db.open() {
            DataStorageUtils.error(errorMessage, db, call)
            return
        }
        
        if let errorMessage = db.deleteRow(key: key) {
            DataStorageUtils.error(errorMessage, db, call)
        } else {
            DataStorageUtils.success([:], db, call)
        }
    }
    
    @objc func drop(_ call: CAPPluginCall){
        guard let table = call.getString("table") else {
            call.error(DataStorageError.EmptyTable)
            return
        }
        
        let db = SqliteDB(dbPath: self.dbName, tableName: table)
        
        if !db.existDatabase() {
            call.error(DataStorageError.DatabaseNotFound)
            return
        }
        
        if let errorMessage = db.open() {
            DataStorageUtils.error(errorMessage, db, call)
            return
        }
        
        if let errorMessage = db.dropTable() {
            DataStorageUtils.error(errorMessage, db, call)
        } else {
            DataStorageUtils.success([:], db, call)
        }
    }
    
    @objc func remove(_ call: CAPPluginCall) {
        let db = SqliteDB(dbPath: self.dbName, tableName: "")
        
        if !db.existDatabase() {
            call.error(DataStorageError.DatabaseNotFound)
            return
        }
        
        if let errorMessage = db.removeDatabase() {
            DataStorageUtils.error(errorMessage, db, call)
        } else {
            DataStorageUtils.success([:], db, call)
        }
    }
    
    @objc func retrieve(_ call: CAPPluginCall) {
        guard let table = call.getString("table") else {
            call.error(DataStorageError.EmptyTable)
            return
        }
        
        guard let key = call.getString("key") else {
            call.error(DataStorageError.EmptyKey)
            return
        }
        
        let db = SqliteDB(dbPath: self.dbName, tableName: table)
        
        if !db.existDatabase() {
            call.error(DataStorageError.DatabaseNotFound)
            return
        }
        
        if let errorMessage = db.open() {
            DataStorageUtils.error(errorMessage, db, call)
            return
        }
        
        let storedValue = db.selectOne(key: key)
        print("DATA STORAGE -> result: \(storedValue)")
        if !storedValue.starts(with: "{") {
            DataStorageUtils.error(storedValue, db, call)
            return
        }
        
        guard let value = DataStorageUtils.jsonParse(storedValue) else {
            DataStorageUtils.error(DataStorageError.JsonParse, db, call)
            return
        }
        DataStorageUtils.success(value, db, call)
        
    }
    
    @objc func retrieveAll(_ call: CAPPluginCall) {
        guard let table = call.getString("table") else {
            call.error(DataStorageError.EmptyTable)
            return
        }
        
        let db = SqliteDB(dbPath: self.dbName, tableName: table)
        
        if !db.existDatabase() {
            call.error(DataStorageError.DatabaseNotFound)
            return
        }
        
        if let errorMessage = db.open() {
            DataStorageUtils.error(errorMessage, db, call)
            return
        }
        
        let storedKeyValues = db.selectAll()
        if storedKeyValues.index(forKey: DataStorageUtils.DATA_STORAGE_ERROR_HASH) != nil {
            let selectAllErrorMessage = storedKeyValues[DataStorageUtils.DATA_STORAGE_ERROR_HASH]
            DataStorageUtils.error(selectAllErrorMessage as! String, db, call)
            return
        }
    
        DataStorageUtils.success(storedKeyValues, db, call)
    }
    
    @objc func store(_ call: CAPPluginCall) {
        // TODO: Add type to avoid problems with boolean and numbers
        // TODO: Now transforms 0 to false.
        
        guard let table = DataStorageUtils.getNoEmptyString("table", call) else {
            call.error(DataStorageError.EmptyTable)
            return
        }
        
        guard let key =  DataStorageUtils.getNoEmptyString("key", call) else {
            call.error(DataStorageError.EmptyKey)
            return
        }
        
        guard let objectValue: [String:Any] = DataStorageUtils.getObjectToStore("value", call) else {
            call.error(DataStorageError.EmptyValue)
            return
        }
        
        guard let value: String = DataStorageUtils.jsonStringify(json: objectValue) else {
            call.error(DataStorageError.JsonStringify)
            return
        }
        
        let db = SqliteDB(dbPath: self.dbName, tableName: table)
        if let errorMessage = db.open() {
            DataStorageUtils.error(errorMessage, db, call)
            return
        }
        
        if let errorMessage = db.createTable() {
            DataStorageUtils.error(errorMessage, db, call)
        } else {
            if let errorMessage = db.insert(key: key, value: value) {
                DataStorageUtils.error(errorMessage, db, call)
            } else {
                DataStorageUtils.success([:], db, call)
            }
        }
    }
    
}

File: capacitor/plugins-library/android/src/main/java/com/krikoo/capacitor/datastorage/DataStorage.java
package com.krikoo.capacitor.datastorage;

import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;

import com.krikoo.capacitor.datastorage.models.DataStorageError;
import com.krikoo.capacitor.datastorage.models.SqliteDB;

@NativePlugin()
public class DataStorage extends Plugin {

  private String dbName = "datastorage.sqlite";

  @PluginMethod()
  public void database(PluginCall call) {
    String name = call.getString("name");
    if (name == null) {
      call.error(DataStorageError.EmptyDatabaseName);
      return;
    }
    this.dbName = String.format("%s.sqlite", name);
    call.success(new JSObject());
  }

  @PluginMethod()
  public void delete(PluginCall call) {
    String table = DataStorageUtils.getNoEmptyString("table", call);
    if (table == null) {
      call.error(DataStorageError.EmptyTable);
      return;
    }

    String key = DataStorageUtils.getNoEmptyString("key", call);
    if (key == null) {
      call.error(DataStorageError.EmptyKey);
      return;
    }

    SqliteDB db = new SqliteDB(this.dbName, table);

    if (!db.existDatabase(bridge)) {
      call.error(DataStorageError.DatabaseNotFound);
      return;
    }

    String errorMessage = db.open(bridge);
    if (errorMessage != null) {
      DataStorageUtils.error(errorMessage, db, call);
      return;
    }

    String deleteErrorMessage = db.deleteRow(key);
    if (deleteErrorMessage != null) {
      DataStorageUtils.error(deleteErrorMessage, db, call);
    } else {
      DataStorageUtils.success(new JSObject(), db, call);
    }
  }

  @PluginMethod()
  public void drop(PluginCall call) {
    String table = DataStorageUtils.getNoEmptyString("table", call);
    if (table == null) {
      call.error(DataStorageError.EmptyTable);
      return;
    }

    SqliteDB db = new SqliteDB(this.dbName, table);

    if (!db.existDatabase(bridge)) {
      call.error(DataStorageError.DatabaseNotFound);
      return;
    }

    String errorMessage = db.open(bridge);
    if (errorMessage != null) {
      DataStorageUtils.error(errorMessage, db, call);
      return;
    }

    String dropErrorMessage = db.dropTable();
    if (dropErrorMessage != null) {
      DataStorageUtils.error(dropErrorMessage, db, call);
    } else {
      DataStorageUtils.success(new JSObject(), db, call);
    }
  }

  @PluginMethod()
  public void remove(PluginCall call) {
    SqliteDB db = new SqliteDB(this.dbName, "");

    if (!db.existDatabase(bridge)) {
      call.error(DataStorageError.DatabaseNotFound);
      return;
    }

    String errorMessage = db.removeDatabase(bridge);
    if (errorMessage != null) {
      call.error(errorMessage);
    } else {
      call.success(new JSObject());
    }
  }

  @PluginMethod()
  public void retrieve(PluginCall call) {
    String table = DataStorageUtils.getNoEmptyString("table", call);
    if (table == null) {
      call.error(DataStorageError.EmptyTable);
      return;
    }

    String key = DataStorageUtils.getNoEmptyString("key", call);
    if (key == null) {
      call.error(DataStorageError.EmptyKey);
      return;
    }

    SqliteDB db = new SqliteDB(this.dbName, table);

    if (!db.existDatabase(bridge)) {
      call.error(DataStorageError.DatabaseNotFound);
      return;
    }

    String errorMessage = db.open(bridge);
    if (errorMessage != null) {
      DataStorageUtils.error(errorMessage, db, call);
      return;
    }

    String storedValue = db.selectOne(key);
    if (!storedValue.startsWith("{")) {
      DataStorageUtils.error(storedValue, db, call);
    } else {
      JSObject value = DataStorageUtils.jsonParse(storedValue);
      if (value == null) {
        DataStorageUtils.error(DataStorageError.JsonParse, db, call);
      } else {
        DataStorageUtils.success(value, db, call);
      }
    }
  }

  @PluginMethod()
  public void retrieveAll(PluginCall call) {
    String table = DataStorageUtils.getNoEmptyString("table", call);
    if (table == null) {
      call.error(DataStorageError.EmptyTable);
      return;
    }

    SqliteDB db = new SqliteDB(this.dbName, table);

    if (!db.existDatabase(bridge)) {
      call.error(DataStorageError.DatabaseNotFound);
      return;
    }

    String errorMessage = db.open(bridge);
    if (errorMessage != null) {
      DataStorageUtils.error(errorMessage, db, call);
      return;
    }

    JSObject storedKeyValues = db.selectAll();
    String selectAllErrorMessage = storedKeyValues.getString(DataStorageUtils.DATA_STORAGE_ERROR_HASH, null);
    if (selectAllErrorMessage != null) {
      DataStorageUtils.error(selectAllErrorMessage, db, call);
    }

    DataStorageUtils.success(storedKeyValues, db, call);
  }

  @PluginMethod()
  public void store(PluginCall call) {
    String table = DataStorageUtils.getNoEmptyString("table", call);
    if (table == null) {
      call.error(DataStorageError.EmptyTable);
      return;
    }

    String key = DataStorageUtils.getNoEmptyString("key", call);
    if (key == null) {
      call.error(DataStorageError.EmptyKey);
      return;
    }

    JSObject objectValue = DataStorageUtils.getObjectToStore("value", call);
    if (objectValue == null) {
      call.error(DataStorageError.EmptyValue);
      return;
    }

    String value = objectValue.toString();

    SqliteDB db = new SqliteDB(this.dbName, table);
    String errorMessage = db.open(bridge);
    if (errorMessage != null) {
      DataStorageUtils.error(errorMessage, db, call);
      return;
    }

    String createErrorMessage = db.createTable();
    if (createErrorMessage != null) {
      DataStorageUtils.error(createErrorMessage, db, call);
    } else {
      String insertErrorMessage = db.insert(key, value);
      if (insertErrorMessage != null) {
        DataStorageUtils.error(insertErrorMessage, db, call);
      } else {
        DataStorageUtils.success(new JSObject(), db, call);
      }
    }
  }

}
